pyomo/util/subsystems.py

```diff
@@ -0,0 +1,195 @@
+                   subsystem blocks.
+    fix_inputs: Whether to fix variables in the constraints that are not
+                included in the subsystem.
+
+    Yields
+    ------
+    "Subsystem blocks" containing the variables and constraints specified
+    by each entry in subsystems. Variables in the constraints that are
+    not specified are contained in the input_vars component.
+    """
+    for cons, vars in subsystems:
+        block = create_subsystem_block(cons, vars, include_fixed)
+        if fix_inputs:
+            to_fix = list(block.input_vars[:])
+            with TemporarySubsystemManager(to_fix=to_fix):
+                # On enter, we fix the "inputs" of the subsystem. When
+                # control returns to this code (iterator advance), we
+                # exit this context, and inputs are unfixed.
+                yield block
+        else:
+            yield block
+
+
+class TemporarySubsystemManager(object):
+    """ This class is a context manager for cases when we want to
+    temporarily fix or deactivate certain variables or constraints
+    in order to perform some solve or calculation with the resulting
+    subsystem.
+
+    """
+
+    def __init__(self, to_fix=None, to_deactivate=None, to_reset=None):
+        if to_fix is None:
+            to_fix = []
+        if to_deactivate is None:
+            to_deactivate = []
+        if to_reset is None:
+            to_reset = []
+        self._vars_to_fix = to_fix
+        self._cons_to_deactivate = to_deactivate
+        self._comps_to_set = to_reset
+        self._var_was_fixed = None
+        self._con_was_active = None
+        self._comp_original_value = None
+
+    def __enter__(self):
+        to_fix = self._vars_to_fix
+        to_deactivate = self._cons_to_deactivate
+        to_set = self._comps_to_set
+        self._var_was_fixed = [(var, var.fixed) for var in to_fix]
+        self._con_was_active = [(con, con.active) for con in to_deactivate]
+        self._comp_original_value = [(comp, comp.value) for comp in to_set]
+
+        for var in self._vars_to_fix:
+            var.fix()
+
+        for con in self._cons_to_deactivate:
+            con.deactivate()
+
+        return self
+
+    def __exit__(self, ex_type, ex_val, ex_bt):
+        for var, was_fixed in self._var_was_fixed:
+            if not was_fixed:
+                var.unfix()
+        for con, was_active in self._con_was_active:
+            if was_active:
+                con.activate()
+        for comp, val in self._comp_original_value:
+            comp.set_value(val)
+
+
+class ParamSweeper(TemporarySubsystemManager):
+    """ This class enables setting values of variables/parameters
+    according to a provided sequence. Iterating over this object
+    sets values to the next in the sequence, at which point a
+    calculation may be performed and output values compared.
+    On exit, original values are restored.
+    """
```

Comment:
I am a bit confused on the use of this class:
- It is named "ParamSweeper", but it seems to be focused on Vars (and not Params - can it be used with Params, too?).
- What is the distinction between "input" and "output"?  It seems like they are all just "values"

I think the documentation needs an example: I had to go to the tests to try and understand the intended functionality.

Reply:
- Yes, mutable parameters may be included in `input_values`. It has additional functionality to handle the case when variables are used as "parameters" or "inputs" into a block of equations.
- The distinction is that inputs are set every time this iterator is advanced. Outputs are presumably solved for (by the user) in this object's context manager. Their expected values are provided by the user in a convenient data structure (component map from vars to lists), who can then write one loop to check actual vs. expected values using the data structure returned by the iterator (which only contains the values for the current scenario and is therefore hopefully more convenient).

Hopefully that did not make this too much more confusing. I also have added an example to the docstring.